# Assignment 2 _ 05.09.23 _ Batch:B _ Group:9

In [2]:
import re

### 1) For any input string write a python function using regular expressions to        
#### i.Extract email id.                                                                                                             
#### ii.Extract string by removing mail id.

In [4]:
def extract_and_remove_emails(input_string):
    email_pattern = r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,7}\b'
    email_list = re.findall(email_pattern, input_string)
    clean_string = re.sub(email_pattern, '', input_string)

    return email_list, clean_string

input_string = "Please contact john.doe@example.com for more information. You can also reach out to jane.smith@example.org."
emails, cleaned_text = extract_and_remove_emails(input_string)
print("Extracted Email IDs:", emails)
print("Cleaned Text:", cleaned_text)


Extracted Email IDs: ['john.doe@example.com', 'jane.smith@example.org']
Cleaned Text: Please contact  for more information. You can also reach out to .


1. Use the `re.findall` function to find all non-overlapping matches of the `email_pattern` in the input string. This will extract all email addresses from the input string and store them in the `email_list`.

2. Use the `re.sub` function to replace all matches of the `email_pattern` in the input string with an empty string, effectively removing the email addresses. The result is stored in the `clean_string`.

3. Return both the `email_list` (list of extracted email addresses) and the `clean_string` (input string with email addresses removed) as a tuple.

# ********************************************************************************************

### 2) Write a python function using regular expressions to find the count of vowels in an input string

In [4]:
def count_vowels(input_string):
    pattern = r'[aeiouAEIOU]'
    vowels = re.findall(pattern, input_string)
    return len(vowels)

user_input = input("Enter a string: ")
vowel_count = count_vowels(user_input)
print("Number of vowels in the input string:", vowel_count)

Enter a string: My name is Shruthi
Number of vowels in the input string: 5


1. First we define a regular expression pattern using the `re` module. The pattern `[aeiouAEIOU]` matches any vowel in both lowercase and uppercase letters. So, this pattern will match all vowels in the input string.

2. The `re.findall()` function is used to find all occurrences of the vowel pattern in the `input_string`. It returns a list of all the matched vowels in the input string.

3. Then the length of the list `vowels`,is recorded.

# ********************************************************************************************

### 3) Write a python function using regular expressions to check if a given string,is palindrome. 

In [5]:
def is_palindrome(input_string):
    cleaned_string = re.sub(r'[^a-zA-Z0-9]', '', input_string).lower()
    return cleaned_string == cleaned_string[::-1]

user_input = input("Enter a string: ")
if is_palindrome(user_input):
    print("The input string is a palindrome.")
else:
    print("The input string is not a palindrome.")

Enter a string: Madam
The input string is a palindrome.


The `is_palindrome` function:
   - It takes an `input_string` as its argument.
   - `re.sub(r'[^a-zA-Z0-9]', '', input_string)` uses regular expressions to remove all characters from `input_string` that are not alphanumeric.
   - `.lower()` converts the resulting cleaned string to lowercase, making the palindrome check case-insensitive.
   - `cleaned_string[::-1]` is used to reverse the cleaned string.
   - The function returns `True` if the cleaned string is the same as its reverse i.e., it's a palindrome.


# ********************************************************************************************

### 4)What will be the output of the following Python code? Explain each line of code in detail.
#### import re s = "A new day" 
#### m = re.match(r'(.*)(.*?)', s)
#### print(m.group(2)) print(m.group(0))

In [7]:
import re
s = "A new day"
m = re.match(r'(.)(.?)', s)
print(m.group(2)) 
print(m.group(0))

 
A 


We first import the re module (regular expressions)
We then initialize a string variable s with the value "A new day".

Furthur we match the regular expression pattern provided as the first argument (r'(.)(?)') against the input string s.

r'(.)(?)': This pattern consists of two capturing groups:

-(.+): The first group (.) captures any single character and is represented by (.+).

-(?): The second group (.) captures an optional single character and is represented by (?).

The re.match() function returns a match object (m) if the pattern is found at the beginning of the string. In this case, it will match the first two characters, which are "A ".

print(m.group(2)): This line attempts to print the content of the second capturing group ((?)) from the match object m. However, since the second capturing group (?) is optional, it will match an empty string if it doesn't exist. In this case, it will print an empty string.

print(m.group(0)): This line attempts to print the entire match (the entire string that matched the regular expression pattern) from the match object m. In this case, it will print "A ", which is the first two characters of the input string "A new day".

# ********************************************************************************************

### 5) What will be the output of the following Python code? Explain each line of code in detail. 
#### p = re.compile('hello') 
#### r = p.match('hello everyone') 
#### print(r.group(0)) 
#### r = re.match('hello', 'hello everyone') 
#### print(r.group(0))

In [8]:
p = re.compile('hello')
r = p.match('hello everyone')
print(r.group(0))
r = re.match('hello', 'hello everyone')
print(r.group(0))

hello
hello


The first line compiles a regular expression pattern `'hello'` using the `re.compile()` function and assigns the compiled pattern to the variable `p`. The pattern `'hello'` represents the string "hello" that we want to match in the input.

   - `p.match('hello everyone')`: This code uses the compiled regular expression `p` to perform a match on the input string `'hello everyone'`.
   - The `match()` method checks if the pattern `'hello'` matches at the beginning of the input string. In this case, it does match because "hello" appears at the beginning of the string.
   - The match object `r` is assigned the result of the match.

 The `print(r.group(0))` line attempts to print the content of the first capturing group in the match object `r`. In this case, there are no capturing groups specified in the regular expression pattern, so `r.group(0)` represents the entire match.

 The `r = re.match('hello', 'hello everyone')`line uses the `re.match()` function directly to match the pattern `'hello'` against the input string `'hello everyone'`. It does not involve the compiled pattern `p`.

Similarly the `print(r.group(0))` statement, this line attempts to print the content of the first (and only) capturing group in the match object `r` returned by the second `re.match()` call.

For the first part (`r = p.match('hello everyone')`):
- The regular expression `'hello'` matches the string "hello" at the beginning of the input, so `r` contains a match object.
- `r.group(0)` prints the entire matched substring, which is "hello."

For the second part (`r = re.match('hello', 'hello everyone')`):
- Here, the `re.match()` function is used directly with the pattern `'hello'` and the input string `'hello everyone'`.
- Again, the regular expression matches the string "hello" at the beginning of the input.
- `r.group(0)` also prints the entire matched substring, which is "hello."

# ********************************************************************************************

### 6)Explain re.compile using an example. Also using python code explain its usage.

re.compile() is a method in Python's re module that allows you to pre-compile a regular expression pattern into a regular expression object. This can be useful when you need to use the same regular expression pattern multiple times, as it can improve performance by avoiding recompilation of the pattern each time you use it. It also makes your code more readable and maintainable

In [11]:
pattern = re.compile(r'\d{2,4}')  
input_str1 = "12345"
input_str2 = "abc123"
input_str3 = "42 is the answer"

match1 = pattern.search(input_str1)
match2 = pattern.search(input_str2)
match3 = pattern.search(input_str3)

if match1:
    print("Matched:", match1.group())
else:
    print("No match found in input_str1")

if match2:
    print("Matched:", match2.group())
else:
    print("No match found in input_str2")

if match3:
    print("Matched:", match3.group())
else:
    print("No match found in input_str3")

Matched: 1234
Matched: 123
Matched: 42


# ********************************************************************************************

### 7)Write a Python program that matches a string that has an ‘a’ followed by zero or more b's. 

In [10]:
def match_pattern(input_string):
    pattern = r'ab*'

    if re.search(pattern, input_string):
        return True
    else:
        return False

user_input = input("Enter a string: ")

if match_pattern(user_input):
    print("The input matches the pattern 'ab*'")
else:
    print("The input does not match the pattern 'ab*'")

Enter a string: qaab
The input matches the pattern 'ab*'


1. Inside the `match_pattern` function, a regular expression pattern `ab*` is defined using the `r` prefix before the string. This pattern means:
   - `a`: Match the character 'a' exactly.
   - `b*`: Match zero or more occurrences of the character 'b'. The asterisk (*) is a quantifier that means "zero or more."

2. The code then uses the `re.search()` function to search for the pattern within the `input_string`. 

3. If `re.search()` finds a match, the `match_pattern` function returns `True`.

# ********************************************************************************************

### 8)Write a Python program to find sequences of lowercase letters joined with an underscore.

In [17]:
def find_lowercase_sequences_with_underscore(input_string):
    pattern = r'[a-z]+_+'
    sequences = re.findall(pattern, input_string)
    return sequences


user_input = input("Enter a string: ")

result = find_lowercase_sequences_with_underscore(user_input)
if result:
    print("Sequences of lowercase letters joined with underscores:")
    for sequence in result:
        print(sequence)
else:
    print("No matching sequences found.")

Enter a string: AQq_EWSws_
Sequences of lowercase letters joined with underscores:
q_
ws_


Inside the find_lowercase_sequences_with_underscore function:
A regular expression pattern r'[a-z]+_+' is defined, which matches sequences of one or more lowercase letters followed by one or more underscores.
It uses re.findall() to find all non-overlapping matches of this pattern in the input_string.
It returns a list of all the matching sequences found in the input string.

# ********************************************************************************************

### 9)Write a Python program to match a string that contains only upper and lowercase letters, numbers, and underscores.

In [20]:
def is_valid_string(input_string):
    pattern = r'^[a-zA-Z0-9_]$'
    return bool(re.match(pattern, input_string))

user_input = input("Enter a string: ")
if is_valid_string(user_input):
    print("The input string is valid.")
else:
    print("The input string is not valid.")

Enter a string: niejns#d
The input string is not valid.


In the function is_valid_string:
1. We first define a regular expression pattern and assigns it to the variable `pattern`. The pattern `^[a-zA-Z0-9_]$` checks whether a string contains exactly one character that is either a letter, a digit, or an underscore. It uses `^` and `$` to ensure that the pattern matches the entire string, not just a part of it.

2.We use the `re.match()` function from the `re` module to check if the `input_string` matches the `pattern`. If it does, it returns `True`; otherwise, it returns `False`. The `bool()` function is used to convert the result of `re.match()` into a Boolean value.

# ********************************************************************************************

### 10) Write a Python program to remove leading zeros from an IP address.

In [12]:
def remove_leading_zeros(ip_address):
    octets = ip_address.split('.')
    cleaned_octets = []

    for octet in octets:
        cleaned_octet = str(int(octet))  
        cleaned_octets.append(cleaned_octet)

    cleaned_ip_address = '.'.join(cleaned_octets)
    return cleaned_ip_address

user_input = input("Enter an IP address: ")
cleaned_ip = remove_leading_zeros(user_input)
print("Cleaned IP address:", cleaned_ip)

Enter an IP address: 192.001.010.002
Cleaned IP address: 192.1.10.2


The `remove_leading_zeros` function:

1. splits the input `ip_address` into octets using the `split` method. The `.` character is used as the separator, so the IP address is divided into a list of strings, where each string represents an octet. For example, "192.168.001.001" would be split into the list `['192', '168', '001', '001']`.

2. The code then initializes an empty list called `cleaned_octets` to store the cleaned octets without leading zeros.

3. It enters a loop to process each octet one by one.

4. Inside the loop, it first converts the current octet (which is a string) into an integer using the `int` function. This conversion will remove any leading zeros from the octet. For example, "001" becomes 1.

5. It then converts the integer back to a string using the `str` function. This step ensures that all octets have the same format, with leading zeros removed. For example, "001" becomes "1".

6. The cleaned octet is appended to the `cleaned_octets` list.

7. After processing all the octets in the input IP address, the code joins the cleaned octets back together into a single string using the `join` method, with a period (`.`) as the separator. This reconstructs the cleaned IP address.

# ********************************************************************************************

### 11) Write a Python program to search a literals string in a string and also find the location within the original string where the pattern occurs 
#### Sample text : 'The quick brown fox jumps over the lazy dog.'
#### Searched words : 'fox'

In [13]:
sample_text = 'The quick brown fox jumps over the lazy dog.'
searched_word = 'fox'
location = sample_text.find(searched_word)

if location != -1:
    print(f"The word '{searched_word}' was found at index {location}.")
else:
    print(f"The word '{searched_word}' was not found in the sample text.")

The word 'fox' was found at index 16.


1. `sample_text` is a string containing the text in which you want to search for a specific word.
2. `searched_word` is the word you want to find within the `sample_text`.
3. `location` is a variable that stores the result of searching for `searched_word` within `sample_text` using the `find()` method. The `find()` method returns the index (position) of the first occurrence of the specified word within the string. If the word is not found, it returns -1.
4. The code checks whether `location` is not equal to -1. If it's not equal to -1, it means that the `searched_word` was found in the `sample_text`. In this case, it prints a message indicating that the word was found and displays the index (position) at which it was found.
5. If `location` is equal to -1, it means that the `searched_word` was not found in the `sample_text`. In this case, it prints a message indicating that the word was not found.

# ********************************************************************************************

### 12)Write a Python program to convert a date of yyyy-mm-dd format to dd-mm-yyyy format.

In [14]:
def convert_date_format(input_date):
    date_pattern = r'(\d{4})-(\d{2})-(\d{2})'
    formatted_date = re.sub(date_pattern, r'\3-\2-\1', input_date)
    return formatted_date

input_date =input ("Enter date(in the form yyyy-mm-dd):")
formatted_date = convert_date_format(input_date)
print("Input Date:", input_date)
print("Formatted Date:", formatted_date)

Enter date(in the form yyyy-mm-dd):2023-10-04
Input Date: 2023-10-04
Formatted Date: 04-10-2023


The `convert_date_format` function:
   - Define a regular expression pattern called `date_pattern`. This pattern is designed to match dates in the format 'YYYY-MM-DD', where 'YYYY' represents the year, 'MM' represents the month, and 'DD' represents the day, each consisting of two digits.

   - Use the `re.sub` function to replace the matched date pattern with a new string. In this case, it replaces the 'YYYY-MM-DD' format with 'DD-MM-YYYY' format, effectively swapping the day and year portions of the date.

   - Store the result in the `formatted_date` variable.



# ********************************************************************************************